MoreLinq.Test/PartitionTest.cs

```diff
@@ -0,0 +1,194 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2008 Jonathan Skeet. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+using System.Linq;
+using NUnit.Framework;
+
+namespace MoreLinq.Test
+{
+    using System;
+
+    [TestFixture]
+    public class PartitionTest
+    {
+        [Test]
+        public void PartitionWithResultSelector()
+        {
+            var (evens, odds) =
+                Enumerable.Range(0, 10)
+                          .Partition(x => x % 2 == 0, ValueTuple.Create);
+
+            Assert.That(evens, Is.EquivalentTo(new[] { 0, 2, 4, 6, 8 }));
+            Assert.That(odds,  Is.EquivalentTo(new[] { 1, 3, 5, 7, 9 }));
+        }
+
+        [Test]
+        public void PartitionBooleanGrouping()
+        {
+            var (evens, odds) =
+                Enumerable.Range(0, 10)
+                          .GroupBy(x => x % 2 == 0)
+                          .Partition(ValueTuple.Create);
+
+            Assert.That(evens, Is.EquivalentTo(new[] { 0, 2, 4, 6, 8 }));
+            Assert.That(odds,  Is.EquivalentTo(new[] { 1, 3, 5, 7, 9 }));
+        }
+
+        [Test]
+        public void PartitionNullableBooleanGrouping()
+        {
+            var xs = new int?[] { 1, 2, 3, null, 5, 6, 7, null, 9, 10 };
+
+            var (lt5, gte5, nils) =
+                xs.GroupBy(x => x != null ? x < 5 : (bool?) null)
+                  .Partition(ValueTuple.Create);
+
+            Assert.That(lt5,  Is.EquivalentTo(new[] { 1, 2, 3 }));
+            Assert.That(gte5, Is.EquivalentTo(new[] { 5, 6, 7, 9, 10 }));
+            Assert.That(nils, Is.EquivalentTo(new int?[] { null, null }));
+        }
+
+        [Test]
+        public void PartitionBooleanGroupingWithSingleKey()
+        {
+            var (m3, etc) =
+                Enumerable.Range(0, 10)
+                          .GroupBy(x => x % 3)
+                          .Partition(0, ValueTuple.Create);
+
+            Assert.That(m3, Is.EquivalentTo(new[] { 0, 3, 6, 9 }));
+
+            using (var r = etc.Read())
+            {
+                var r1 = r.Read();
+                Assert.That(r1.Key, Is.EqualTo(1));
+                Assert.That(r1, Is.EquivalentTo(new[] { 1, 4, 7 }));
+
+                var r2 = r.Read();
+                Assert.That(r2.Key, Is.EqualTo(2));
+                Assert.That(r2, Is.EquivalentTo(new[] { 2, 5, 8 }));
```

Comment:
This test (and all the tests that have more than one grouping in `etc`) assert that the result is ordered by the order in which keys are seen, but that is not documented. So either the documentation should be updated to include this guarantee, or change these tests to relax that requirement.